reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunity.java

```diff
@@ -16,27 +16,120 @@
+    if (this.shaderJob.getPipelineInfo().hasUniform(arrayName)) {
+      values = this.shaderJob.getPipelineInfo().getArgs(arrayName);
+      if (values.contains(((IntConstantExpr)this.literalExpr).getNumericValue())) {
+        index = values.indexOf(((IntConstantExpr)this.literalExpr).getNumericValue());
+      } else {
+        index = this.shaderJob.getPipelineInfo().getArgs(arrayName).size();
+        values.add(((IntConstantExpr)this.literalExpr).getNumericValue());
+      }
+
+    } else {
+      values = new ArrayList<>();
+      values.add(((IntConstantExpr)this.literalExpr).getNumericValue());
+      index = values.size() - 1;
+    }
+
+    // Removes the old uniform array and adds the new one to the associated json structure.
+    this.shaderJob.getPipelineInfo().removeUniform(arrayName);
+    final int binding = this.shaderJob.getPipelineInfo().getNumUniforms();
+    this.shaderJob.getPipelineInfo().addUniform(arrayName, BasicType.INT,
+        Optional.of(values.size()), values);
+    this.shaderJob.getPipelineInfo().addUniformBinding(arrayName, false, binding);
+
+    // Adds declaration for the uniform array and replaces the literal with an element in the array.
+    for (TranslationUnit tu: shaderJob.getShaders()) {
+
+      // Declares the array if it didn't exist. Otherwise, removes it and adds a new declaration
+      // with increased size.
+      Optional<Declaration> oldDeclaration = Optional.empty();
+      final List<Declaration> declarations = tu.getTopLevelDeclarations();
+      for (Declaration declaration : declarations) {
+        if (declaration instanceof InterfaceBlock) {
+          final List<String> members = ((InterfaceBlock) declaration).getMemberNames();
+          if (members.get(0).equals(arrayName)) {
+            oldDeclaration = Optional.of(declaration);
+          }
+        }
+      }
+
+      oldDeclaration.ifPresent(tu::removeTopLevelDeclaration);
+
+      final ArrayType arraySize = new ArrayType(BasicType.INT, new ArrayInfo(
+          new IntConstantExpr(String.valueOf(values.size()))));
+
+      arraySize.getArrayInfo().setConstantSizeExpr(2);
+      final InterfaceBlock ib =
+          new InterfaceBlock(Optional.of(new LayoutQualifierSequence(new SetLayoutQualifier(0),
+              new BindingLayoutQualifier(binding))),
+              TypeQualifier.UNIFORM,
+              "buf" + binding,
+              Collections.singletonList(arrayName),
+              Collections.singletonList(arraySize),
+              Optional.empty()
+          );
+
+      tu.addDeclaration(ib);
+
+      // Replaces the literal with an element in the uniform array.
+      final IParentMap parentMap = IParentMap.createParentMap(tu);
+      final ArrayIndexExpr aie = new ArrayIndexExpr(new VariableIdentifierExpr(arrayName),
+          new IntConstantExpr(String.valueOf(index)));
+
+      parentMap.getParent(this.literalExpr).replaceChild(this.literalExpr,
+          aie);
```

Comment:
A problem here is that the literal expression will not appear in *every* translation unit associated with the shader job, but only in one of them.

I therefore think that you should record the specific translation unit in which the literal appears as a field of this opportunity class, in addition to the shader job.  Then you should move this code to after the loop that goes through all the translation units, and have it do the parentmap thing to just the translation unit in which the literal expression exists.

I believe in its current form you would find you get an exception if you try this opportunity on a multi-shader shader job - it would be good to have a test for that.

Reply:
Works with multi-shader shader job now.